src/coreclr/jit/valuenum.cpp

```diff
@@ -8492,7 +8492,98 @@ void Compiler::fgValueNumberSsaVarDef(GenTreeLclVarCommon* lcl)
     }
 }
 
-void Compiler::fgValueNumberTree(GenTree* tree)
+//------------------------------------------------------------------------
+// fgValueNumberConstStringElemLoad: Try to match "cns_str"[cns_index] tree
+//    and apply a constant VN representing given char at cns_index in that string.
+//
+// Arguments:
+//    tree - the GT_IND node
+//
+// Return Value:
+//    true if the pattern was recognized and a new VN is assigned
+//
+bool Compiler::fgValueNumberConstStringElemLoad(GenTreeIndir* tree)
+{
+    ValueNum  addrVN = tree->gtGetOp1()->gtVNPair.GetLiberal();
+    VNFuncApp funcApp;
+    if (!varTypeIsShort(tree) || !vnStore->GetVNFunc(addrVN, &funcApp))
```

Comment:
Why are we interested in `TYP_SHORT` (signed) trees, since I assume this is a TP heuristic?

This shouldn't have to recompute `addrVN` + `GetVNFunc`, since the caller already did that.

Reply:
@SingleAccretion can you point me where exactly caller already did it? it seems like other paths also compute it https://github.com/dotnet/runtime/blob/b406c1782d15a676b3e70cf71c53393cbc99481e/src/coreclr/jit/valuenum.cpp#L8838-L8853.

Changed to TYP_SHORT (yes, it was a TP fast-out path)

Reply:
> can you point me where exactly caller already did it? it seems like other paths also compute it

`else if (vnStore->GetVNFunc(addrNvnp.GetLiberal(), &funcApp)` computes it. It would be good to change this to compute the thing once like `ASG` numbering does.

Reply:
do you mind if I leave it as is, I've just pushed a change to extract vnfunc only after TP-oriented checks. I tried to re-organize code to save these two existing lookups and didn't like the outcome, feel free to file a PR to clean it up

Reply:
Sure.